astroquery/eso/utils.py

```diff
@@ -0,0 +1,197 @@
+"""
+utils.py: helper functions for the astropy.eso module
+"""
+from dataclasses import dataclass
+from typing import Dict, List, Optional, Union
+from astropy.table import Table
+
+DEFAULT_LEAD_COLS_RAW = ['object', 'ra', 'dec', 'dp_id', 'date_obs', 'prog_id']
+DEFAULT_LEAD_COLS_PHASE3 = ['target_name', 's_ra', 's_dec', 'dp_id', 'date_obs', 'proposal_id']
+
+
+@dataclass
+class _UserParams:
+    """
+    Parameters set by the user
+    """
+    table_name: str
+    column_name: str = None
+    allowed_values: Union[List[str], str] = None
+    cone_ra: float = None
+    cone_dec: float = None
+    cone_radius: float = None
+    columns: Union[List, str] = None
+    column_filters: Dict[str, str] = None
+    top: int = None
+    order_by: str = ''
+    order_by_desc: bool = True
+    count_only: bool = False
+    query_str_only: bool = False
+    print_help: bool = False
+    authenticated: bool = False
+
+
+def _split_str_as_list_of_str(column_str: str):
+    if column_str == '':
+        column_list = []
+    else:
+        column_list = list(map(lambda x: x.strip(), column_str.split(',')))
+    return column_list
+
+
+def _raise_if_has_deprecated_keys(filters: Optional[Dict[str, str]]) -> bool:
+    if not filters:
+        return
+
+    if any(k in filters for k in ("box", "coord1", "coord2")):
+        raise ValueError(
+            "box, coord1 and coord2 are deprecated; "
+            "use cone_ra, cone_dec and cone_radius instead."
+        )
+
+    if any(k in filters for k in ("etime", "stime")):
+        raise ValueError(
+            "'stime' and 'etime' are deprecated; "
+            "use instead 'exp_start' together with '<', '>', 'between'. Examples:\n"
+            "\tcolumn_filters = {'exp_start': '< 2024-01-01'}\n"
+            "\tcolumn_filters = {'exp_start': '>= 2023-01-01'}\n"
+            "\tcolumn_filters = {'exp_start': \"between '2023-01-01' and '2024-01-01'\"}\n"
+        )
+
+
+def _build_where_constraints(
+        column_name: str,
+        allowed_values: Union[List[str], str],
+        column_filters: Dict[str, str]) -> str:
+    def _format_helper(av):
+        if isinstance(av, str):
+            av = _split_str_as_list_of_str(av)
+        quoted_values = [f"'{v.strip()}'" for v in av]
+        return f"{column_name} in ({', '.join(quoted_values)})"
+
+    column_filters = column_filters or {}
+    where_constraints = []
+    if allowed_values:
+        where_constraints.append(_format_helper(allowed_values))
+
+    where_constraints += [
+        f"{k} {_adql_sanitize_op_val(v)}" for k, v in column_filters.items()
+    ]
+    return where_constraints
+
+
+def reorder_columns(table: Table,
```

Comment:
I have multiple questions here:

- is this really necessary? Why the default column order is not suffucient?
- but if this is necessary, then please just go by suggested way for table column reordering from the astropy docs, "Select or reorder columns" on this page: https://docs.astropy.org/en/stable/table/modify_table.html

Also, minor nitpicking, but all the other functions here are private, why is this one not?

Reply:
Hmmm... but the function is already doing exactly what the docs suggest (isn't it?)...

From the docs:
```
new_order = ['a', 'c', 'b']  # List or tuple
t_acb = t[new_order]
```

From the ESO module:
```
[...]
last_cols = first_cols + last_cols  # this is a list of column names
table = table[last_cols]
```

The code above those two lines is just computing the list of column names.

Other than that,
 - the function is now private, good point.
 - yes, the reordering is necessary, because the default is alphabetical order and the first fields are long urls and other info that is not legible and take much space. The reorder was suggested by several beta testers, which were expecting the columns in the order that this function generates.